File: official/convex.workspace.runtime/implementations/convex_interlocking_infrastructure/fixtures/dirty/interlocking_no_cargo/convex/trains/interlocking.ts
// DIRTY fixture — coder.convex.interlocking-does-not-carry-cargo.
// The InterlockingRunner module references the Cargo data plane and stores an artifact_urn value,
// bleeding the artifact plane into the route-control layer.
// Expected: >=1 violation (interlocking-cargo-mutation).
import { TrainRunner } from "./runner";
import { Cargo } from "./runner";

export interface InterlockingResolution {
  interlockingId: string;
  routeId: string;
  trainId: string;
  trainPath: string;
  category: string;
  categoryDigit: string;
  guardId: string;
  reason: string;
}

export class InterlockingRunner {
  constructor(private readonly path: string) {}

  resolveTrain(action: string, inputs: Record<string, unknown>): InterlockingResolution {
    // BUG: the interlocking reads/mutates Cargo and stores an artifact_urn value mid-resolution.
    const cargo = new Cargo();
    cargo["artifact_urn"] = "urn:artifact:match:result";
    return {
      interlockingId: "interlocking:match-resolution",
      routeId: "nominal-all-voted",
      trainId: "3001-solo-match-complete",
      trainPath: "plan/_trains/3001-solo-match-complete.yaml",
      category: "nominal",
      categoryDigit: "3",
      guardId: "all-voted",
      reason: "all participants voted",
    };
  }

  execute(action: string, inputs: Record<string, unknown>) {
    const resolution = this.resolveTrain(action, inputs);
    return new TrainRunner(resolution.trainPath).execute(resolution.trainId, inputs);
  }
}

File: official/atdd.extension.coder.train-interlocking/implementations/interlocking-infrastructure/fixtures/dirty/python/trains/interlocking.py
"""InterlockingRunner — DIRTY fixture.

A broken route-control layer that competes with TrainRunner and Cargo:
  * resolve_train returns a bare train_id string (no structured InterlockingResolution model)
  * imports a wagon module directly
  * calls run_train(...) directly
  * loops over train.sequence as a step executor
  * references / mutates Cargo and stores artifact_urn values
All forbidden by core afokapu/atdd#1251.
"""
from match.wagon import run_match  # direct wagon import (forbidden)


class InterlockingRunner:
    def __init__(self, interlocking_yaml_path):
        self._path = interlocking_yaml_path

    def resolve_train(self, action, inputs, state=None):
        # BUG: bare train_id string, not a structured resolution model.
        return "3001-solo-match-complete"

    def execute(self, action, inputs, state=None, timing=None, capture_trace=True):
        train = _load_train(self._path)
        cargo = Cargo(inputs)  # references Cargo symbol (forbidden)
        for step in train.sequence:  # loops over train.sequence as executor (forbidden)
            cargo = run_train(step, cargo)  # calls run_train directly (forbidden)
        cargo["artifact_urn:match"] = run_match(cargo)  # mutates cargo + artifact_urn (forbidden)
        return cargo


def run_train(step, cargo):
    return cargo


class Cargo(dict):
    pass


def _load_train(path):
    class _T:
        sequence = ["match"]

    return _T()
